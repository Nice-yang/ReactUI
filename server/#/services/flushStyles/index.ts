import { StyleObject } from '#/common/models/Style';
import { ReplacementBuilder } from '#/utils/ReplacementBuilder';
import { findTemplateStringAtCursor$ } from '#/utils/tsNode';
import * as _ from 'lodash/fp';
import * as R from 'ramda';
import { pipe } from 'rxjs';
import { map } from 'rxjs/operators';
import * as ts from 'typescript';

const stringifyStyles = (styleObject: StyleObject) => {
  return R.pipe(
    R.keys,
    R.map(property => `${property}: ${styleObject[property]};`),
    R.join('\n'),
  )(styleObject);
};

const flushStyles = (styleObject: StyleObject) =>
  pipe(
    findTemplateStringAtCursor$,
    map(cursorNode => {
      const replacementBuilder = new ReplacementBuilder(
        cursorNode.getSourceFile(),
      );

      const styleString = stringifyStyles(styleObject);

      if (ts.isNoSubstitutionTemplateLiteral(cursorNode.template)) {
        replacementBuilder.replaceNodeWithText(
          cursorNode.template,
          '`\n' + styleString + '\n`',
        );
      } else {
        const templateSpansLength = cursorNode.template.templateSpans.length;
        const tail =
          cursorNode.template.templateSpans[templateSpansLength - 1].literal;
        const tailStaticStringMatch = tail.getText().match(/([\w-]*:.*)/);
        const tailStaticStringStartIndex =
          tailStaticStringMatch && tailStaticStringMatch.index
            ? tailStaticStringMatch.index
            : 0;

        const totalLengthWithoutTail =
          cursorNode.template.getText().length - tail.getText().length;

        const tailStaticStringIndexInTemplate =
          totalLengthWithoutTail + tailStaticStringStartIndex;

        const originalTemplateString = _.trimChars(
          '`\n',
          R.slice(0, tailStaticStringIndexInTemplate)(
            cursorNode.template.getText(),
          ),
        );

        replacementBuilder.replaceNodeWithText(
          cursorNode.template,
          '`' + originalTemplateString + styleString + '\n`',
        );
      }

      return replacementBuilder.applyReplacements();
    }),
  );

export default flushStyles;
