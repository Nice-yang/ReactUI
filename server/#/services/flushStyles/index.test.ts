import flushStyles$ from './index';

describe('no substitution template literal', () => {
  const fileNoSubstitutionTemplate1 = `${__dirname}/__fixtures__/noSubstitutionTemplate1.tsx`;

  it('success', done => {
    flushStyles$(
      {
        fileName: fileNoSubstitutionTemplate1,
        lineNumber: 3,
        columnNumber: 13,
      },
      { display: 'flex' },
    ).subscribe(result => {
      expect(result).toMatchSnapshot();
      done();
    });
  });
});

describe('substitution template literal', () => {
  const fileSubstitutionTemplate1 = `${__dirname}/__fixtures__/substitutionTemplate1.tsx`;

  it('should override static string in tail of last template', done => {
    flushStyles$(
      {
        fileName: fileSubstitutionTemplate1,
        lineNumber: 3,
        columnNumber: 13,
      },
      { display: 'flex' },
    ).subscribe(result => {
      expect(result).toMatchSnapshot();
      done();
    });
  });
});
