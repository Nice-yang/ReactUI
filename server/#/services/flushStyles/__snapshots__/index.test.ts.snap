// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`no substitution template literal success 1`] = `
"import * as styled from 'styled';

const Foo = styled.div\`
display: flex;
\`;

export default Foo;
"
`;

exports[`substitution template literal should override static string in tail of last template 1`] = `
"import * as styled from 'styled';

const Foo = styled.div\`  color: red;
  display: \${props => (props.foo ? 'flex' : 'block' && true)};
  \${props => props.bar && 'cursor: pointer;'}
  display: flex;
\`;

export default Foo;
"
`;
